src/Compilers/CSharp/Test/Symbol/Symbols/CovariantReturnTests.cs

```diff
@@ -1659,11 +2115,793 @@ End Class
+  IL_0146:  ldtoken    ""System.Func(Of String)""
+  IL_014b:  call       ""Function System.Type.GetTypeFromHandle(System.RuntimeTypeHandle) As System.Type""
+  IL_0150:  call       ""Function System.Linq.Expressions.Expression.Convert(System.Linq.Expressions.Expression, System.Type) As System.Linq.Expressions.UnaryExpression""
+  IL_0155:  ldc.i4.1
+  IL_0156:  newarr     ""System.Linq.Expressions.ParameterExpression""
+  IL_015b:  dup
+  IL_015c:  ldc.i4.0
+  IL_015d:  ldloc.s    V_15
+  IL_015f:  stelem.ref
+  IL_0160:  call       ""Function System.Linq.Expressions.Expression.Lambda(Of System.Func(Of Derived, System.Func(Of String)))(System.Linq.Expressions.Expression, ParamArray System.Linq.Expressions.ParameterExpression()) As System.Linq.Expressions.Expression(Of System.Func(Of Derived, System.Func(Of String)))""
+  IL_0165:  stloc.s    V_14
+  // sequence point: End Sub
+  IL_0167:  ret
+}
+");
+        }
+
+        [Fact]
+        public void BinaryCompatibility_01()
+        {
+            var s0 = @"
+public class Base
+{
+    public virtual object M() => null;
+}
+";
+            var ref0 = CreateCompilationWithoutCovariantReturns(s0).EmitToImageReference();
+
+            var s1a = @"
+public class Mid : Base
+{
+}
+";
+            var ref1a = CreateCompilationWithoutCovariantReturns(s1a, references: new[] { ref0 }, assemblyName: "ref1").EmitToImageReference();
+
+            var s1b = @"
+public class Mid : Base
+{
+    public override string M() => null;
+}
+";
+            var ref1b = CreateCompilationWithCovariantReturns(s1b, references: new[] { ref0 }, assemblyName: "ref1").EmitToImageReference();
+
+            var s2 = @"
+public class Derived : Mid
+{
+    public override string M() => null;
+}
+";
+            var comp = CreateCompilationWithoutCovariantReturns(s2, references: new[] { ref0, ref1a }).VerifyDiagnostics(
+                // (4,28): error CS8652: The feature 'covariant returns' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //     public override string M() => null;
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "M").WithArguments("covariant returns").WithLocation(4, 28)
+                );
+            verify1(comp);
+
+            comp = CreateCompilationWithCovariantReturns(s2, references: new[] { ref0, ref1a }).VerifyDiagnostics(
+                );
+            verify1(comp);
+            verify1(CompilationReferenceView(comp, ref0, ref1a));
+            verify1(MetadataView(comp, ref0, ref1a));
+            verify1(RetargetedView(comp, ref0, ref1a));
+
+            verify2(CompilationReferenceView(comp, ref0, ref1b));
+            verify2(MetadataView(comp, ref0, ref1b));
+            verify2(RetargetedView(comp, ref0, ref1b));
+
+            static void verify1(CSharpCompilation comp)
+            {
+                VerifyOverride(comp, "Derived.M", "System.String Derived.M()", "System.Object Base.M()");
```

Comment:
>VerifyOverride [](start = 16, length = 14)

In addition to this check it would be good to observe what are we going to call for an invocation of ```d.M()``` for a given set of references, where d is an instance of Derived.

Reply:
This PR is for compile-time behavior only. We already have tests that demonstrate what IL is produced by the compiler for source with particular base classes.

---
In reply to: [424885197](https://github.com/dotnet/roslyn/pull/44025#discussion_r424885197) [](ancestors = 424885197)